### PR TITLE
Bulk update team devices API (bulk move support only)

### LINKED ISSUE #4335: Support for bulk move devices (API)

### PR DESCRIPTION
**Pull request by Steve-Mcl**
closes #4335

## Description

Adds backend API route for team devices bulk update:
* `test/unit/forge/routes/api/teamDevices_spec.js`
* Method:  `PUT`
* URI: `/api/v1/teams/:teamId/devices/bulk`

The route was debated in design phase in the issue but after much debate, I have stuck with the same verb and definition already established in the existing `devices` API (that are responsible for single device combined update / device move endpoint).  More than happy to change this around if desired.

### Tests Added
```
  Team Devices API
    Team Devices
      Supports bulk operations
        update
          move devices
            ✔ Rejects invalid application (404)
            ✔ Rejects invalid instance (404)
            ✔ Rejects moving a device to another team application (400)
            ✔ Rejects moving a device to another team instance (400)
            ✔ Rejects invalid input
            ✔ Member cannot move devices (403)
            ✔ Non team member cannot move devices (404)
            ✔ Move all to Application 2 (48ms)
            ✔ Move all to Project 2 (50ms)
            ✔ Only updates devices not already assigned to the target application (46ms)
            ✔ Only updates devices not already assigned to the target project (49ms)
            ✔ Only unassigns devices not already unassigned
```

## Related Issue(s)

Owner #4335
Parent #4290

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Comment by Steve-Mcl** [comment]
NOTE: I mentioned in Eng Meeting that there was a sensible refactoring - that is the controller method `moveDevices` _should_ replace the code in https://github.com/FlowFuse/flowfuse/blob/f58bc20fb1ffa838dbf24b57240bb91f04eb5955/forge/routes/api/device.js#L435-L518

There reason I was in two minds was related to one of our guys [raising a concern](https://github.com/FlowFuse/flowfuse/issues/4313) about devices being reset when moved (even though they were in developer mode) and whether I would include or exclude that "feature".

As it happens, the implementation of the bulk update is a refactoring of this original route (with obvious modifications for multiple device support) INCLUDING the switching of a device to `"autonomous"` (fleet) mode when unassigned.

As it happens, with the introduction of bulk move feature, the user can side step the unassignment when moving devices from application to application and thus the switching out of developer mode so it kinda becomes  _less of an issue_

<br>
In short, we can and should refactor the single version in the API Route to use the new controller method.

Will raise an issue as a follow up.

File: forge/db/controllers/Device.js [hunk 1]
```diff
@@ -271,5 +271,176 @@ module.exports = {
...
+                    })
+                }
+            } else if (assignTo === 'instance') {
+                // ### Add device to instance ###
+                if (device.ownerType === 'instance' && device.Project?.id === assignToProject.id) {
+                    // Device is already assigned to this instance - nothing to do
+                    continue
+                } else {
+                    device.ProjectId = assignToProject.id
+                    device.ApplicationId = assignToProject.ApplicationId
```
Comment (knolleary) [inline forge/db/controllers/Device.js:370]: When assigning to an instance, `ApplicationId` should be null. My understanding is ProjectId and ApplicationId should never both be set.

**Review by knolleary — CHANGES_REQUESTED** [review]
One issue spotted.

File: forge/db/controllers/Device.js (hunk 1 above)
Comment (Steve-Mcl) [reply forge/db/controllers/Device.js:370]: Hmmmm. very well spotted.

I believe, since the target is being set to the value of an existing/valid source `project` the value is inconsequential and the tests (indirectly) confirm this: https://github.com/FlowFuse/flowfuse/pull/4336/files/05cc75601c14f978b6d5d068c25768c135063863#diff-23b65c9650241a9666db727191beff98fe97e80a3a0175102bebd291692ca6abR709 but for correctness here and now, this should probably be `null`ed

I will switch branch and double check.

ta.

Comment (Steve-Mcl) [reply forge/db/controllers/Device.js]: checked this over. we do need to set this null.

resolved in [f070a56](https://github.com/FlowFuse/flowfuse/pull/4336/commits/f070a56a4f81f17edf4ca6eb1f906be96d411a11)

Thanks.

Comment (knolleary) [reply forge/db/controllers/Device.js]: Conceptually, I was always in favour of setting applicationId when assigned to an instance in that application - as it makes retrieving a list of all devices within the application trivial. But that wasn't the path taken, so yes, we now need to ensure only one or the other is set.

**Review by knolleary — APPROVED** [review]

**Comment by knolleary** [comment]
@Steve-Mcl approved - I appreciate there are a few other PRs behind this. Will let you merge in the appropriate order.

**Comment by Steve-Mcl** [comment]
* All branches merged into this.
* Tested operations on pre-staging
* Once all green - good to go.
